platforms/jvm/plugins-java/src/main/java/org/gradle/api/plugins/JavaBasePlugin.java

```diff
@@ -259,6 +259,8 @@ private void definePathsForSourceSet(final SourceSet sourceSet, final Project pr
     }
 
     private void defineConfigurationsForSourceSet(SourceSet sourceSet, RoleBasedConfigurationContainerInternal configurations) {
+        configurations.recordMaybeCreateContext("sourceSet " + sourceSet.getName() + " setup");
```

Comment:
I realized after seeing your comments that I hadn't made my 💭 comments visible.  Sorry about that.

I also think this situation isn't great as is.  I tried to keep the context variable generic, which makes the configuration container only "aware" of source sets when constructing a warning message in order to provide extra details.  This context could be set to anything other description in other situations where we use `maybeCreate` beyond just here.  This is a very loose connection, but it still is a connection.

I considered:

1. Making a new `MaybeCreateContext` type, with `getAdviceOnInconsistentUsage()`, `getAdditionalDocLink()` etc. methods that the container could call upon failure.  This could isolate the container from the source sets pretty completely at the cost of a little more complexity.
2. Moving the validation logic into the `JavaBasePlugin` itself with a new`maybeCreateConfiguration(String name)` there that would duplicate some of the maybe create logic and test for existence and then create the confs if not found.  I think this is what you are proposing.  I don't like the idea of duplicating so much logic here (because we'd still need maybeCreate in the container).

I was thinking of 1. as mutable state on the container also, and went with the simpler string context.  Now that I think about it more, I think this could be an extra parameter to an overload of `maybeCreate` - this would eliminate mutable state in the container and remove any reference to source sets there.

I'm going to try this.